File: cpp/src/controller.cpp
#include "puyo/controller.hpp"

#include <queue>
#include <tuple>

namespace puyo {

namespace {

constexpr int MAX_ROW = 14;
constexpr int DX[4] = {0, 1, 0, -1};
constexpr int DY[4] = {1, 0, -1, 0};
constexpr int KEY_FRAMES = 2, QUICK_TURN_FRAMES = 4, DROP_FRAMES_PER_ROW = 2, LOCK_FRAMES = 20, TEAR_FRAMES = 20;
constexpr int N_STATES = WIDTH * MAX_ROW * 4;

struct State {
    int x, y, r;
};
int id(const State& s) { return ((s.x - 1) * MAX_ROW + (s.y - 1)) * 4 + s.r; }
State from_id(int i) { return {i / (MAX_ROW * 4) + 1, (i / 4) % MAX_ROW + 1, i % 4}; }

struct Ctl {
    const std::array<int, WIDTH>& h;
    uint8_t top;
    bool free(int x, int y) const {
        if (x < 1 || x > WIDTH || y < 1 || y > MAX_ROW || y <= h[x - 1]) return false;
        return !(y == MAX_ROW && ((top >> (x - 1)) & 1));
    }
    bool valid(int x, int y, int r) const { return free(x, y) && free(x + DX[r], y + DY[r]); }
    bool rotate(const State& s, int d, State& out) const {
        int r2 = (s.r + d + 4) % 4;
        if (valid(s.x, s.y, r2)) return out = {s.x, s.y, r2}, true;
        if (r2 == 1 || r2 == 3) {  // 壁蹴り
            int x2 = s.x - DX[r2];
            if (valid(x2, s.y, r2)) return out = {x2, s.y, r2}, true;
        } else if (r2 == 2) {  // 床蹴り
            if (valid(s.x, s.y + 1, r2)) return out = {s.x, s.y + 1, r2}, true;
        }
        return false;
    }
    bool quick_turn(const State& s, State& out) const {
        State tmp;
        if ((s.r != 0 && s.r != 2) || rotate(s, 1, tmp) || rotate(s, -1, tmp)) return false;
        int r2 = (s.r + 2) % 4;
        if (valid(s.x, s.y, r2)) return out = {s.x, s.y, r2}, true;
        if (r2 == 2 && valid(s.x, s.y + 1, r2)) return out = {s.x, s.y + 1, r2}, true;
        return false;
    }
    // (キー, フレーム, 行き先) を順に返す（Python と同じ順）
    template <class F>
    void neighbors(const State& s, F&& fn) const {
        State t;
        if (valid(s.x - 1, s.y, s.r)) fn('L', KEY_FRAMES, State{s.x - 1, s.y, s.r});
        if (valid(s.x + 1, s.y, s.r)) fn('R', KEY_FRAMES, State{s.x + 1, s.y, s.r});
        if (rotate(s, 1, t)) fn('A', KEY_FRAMES, t);
        if (rotate(s, -1, t)) fn('B', KEY_FRAMES, t);
        if (quick_turn(s, t)) fn('Q', QUICK_TURN_FRAMES, t);
    }
    int drop_frames(const State& s) const {
        int cx = s.x + DX[s.r];
        if (cx == s.x) {
            int low = std::min(s.y, s.y + DY[s.r]);
            return (low - (h[s.x - 1] + 1)) * DROP_FRAMES_PER_ROW + LOCK_FRAMES;
        }
        int ha = h[s.x - 1], hc = h[cx - 1];
        int fall = s.y - (std::max(ha, hc) + 1);
        int tear = ha > hc ? ha - hc : hc - ha;
        int extra = tear ? TEAR_FRAMES + 10 + 6 * (tear - 1) : 0;
        return fall * DROP_FRAMES_PER_ROW + LOCK_FRAMES + extra;
    }
};

const State SPAWN{3, 12, 0};

}  // namespace

Shape shape_of(const Field& f) {
    Shape s;
    for (int x = 1; x <= WIDTH; ++x) {
        s.h[x - 1] = f.height(x);
        if (f.top(x)) s.top |= static_cast<uint8_t>(1u << (x - 1));
    }
    return s;
}

int move_index(Move m) {
    static const auto table = [] {
        std::array<std::array<int, 4>, WIDTH + 2> t{};
        for (auto& row : t) row.fill(-1);
        const auto& all = all_moves();
        for (size_t i = 0; i < all.size(); ++i) t[all[i].x][all[i].rot] = static_cast<int>(i);
        return t;
    }();
    if (m.x < 1 || m.x > WIDTH || m.rot < 0 || m.rot > 3) return -1;
    return table[m.x][m.rot];
}

std::vector<Operation> plan_operations(const Shape& shape) {
    Ctl c{shape.h, shape.top};
    std::vector<Operation> out;
    if (!c.valid(SPAWN.x, SPAWN.y, SPAWN.r)) return out;
    std::vector<int> dist(N_STATES, 1 << 30);
    std::vector<std::string> keys(N_STATES);
    using Item = std::pair<int, int>;
    std::priority_queue<Item, std::vector<Item>, std::greater<Item>> pq;
    dist[id(SPAWN)] = 0;
    pq.emplace(0, id(SPAWN));
    while (!pq.empty()) {
        auto [d, si] = pq.top();
        pq.pop();
        if (d > dist[si]) continue;
        State s = from_id(si);
        c.neighbors(s, [&](char key, int cost, const State& t) {
            int ti = id(t), nd = d + cost;
            std::string cand = keys[si] + key;
            if (nd < dist[ti] || (nd == dist[ti] && cand < keys[ti])) {
                dist[ti] = nd;
                keys[ti] = std::move(cand);
                pq.emplace(nd, ti);
            }
        });
    }
    const auto& all = all_moves();
    std::vector<int> best_frames(all.size(), -1);
    std::vector<std::string> best_keys(all.size());
    for (int i = 0; i < N_STATES; ++i) {
        if (dist[i] >= (1 << 30)) continue;
        State s = from_id(i);
        int mi = move_index(Move{static_cast<int8_t>(s.x), static_cast<int8_t>(s.r)});
        if (mi < 0) continue;
        int total = dist[i] + c.drop_frames(s);
        if (best_frames[mi] < 0 || std::tie(total, keys[i]) < std::tie(best_frames[mi], best_keys[mi])) {
            best_frames[mi] = total;
            best_keys[mi] = keys[i];
        }
    }
    for (size_t i = 0; i < all.size(); ++i)
        if (best_frames[i] >= 0) out.push_back({all[i], best_keys[i], best_frames[i]});
    return out;
}

uint32_t reachable_mask(const Shape& shape) {
    constexpr uint32_t ALL = (1u << 22) - 1;
    int hmax = 0;
    for (int v : shape.h) hmax = std::max(hmax, v);
    if (hmax <= 11) return ALL;  // 12 段以上の列が無ければどこでも置ける

    // 列の高さ（0..13、4bit × 6）と 14 段目（6bit）をキーにした直接マップのキャッシュ
    uint32_t key = shape.top;
    for (int v : shape.h) key = (key << 4) | static_cast<uint32_t>(v);
    struct Entry {
        uint32_t key = ~0u, mask = 0;
    };
    thread_local std::array<Entry, 4096> cache;
    Entry& e = cache[(key * 2654435761u) >> 20];
    if (e.key == key) return e.mask;

    Ctl c{shape.h, shape.top};
    uint32_t mask = 0;
    if (c.valid(SPAWN.x, SPAWN.y, SPAWN.r)) {
        std::array<bool, N_STATES> seen{};
        std::vector<State> stack{SPAWN};
        seen[id(SPAWN)] = true;
        while (!stack.empty()) {
            State s = stack.back();
            stack.pop_back();
            int mi = move_index(Move{static_cast<int8_t>(s.x), static_cast<int8_t>(s.r)});
            if (mi >= 0) mask |= 1u << mi;
            c.neighbors(s, [&](char, int, const State& t) {
                if (!seen[id(t)]) {
                    seen[id(t)] = true;
                    stack.push_back(t);
                }
            });
        }
    }
    e = {key, mask};
    return mask;
}

uint32_t reachable_mask(const Field& f) { return reachable_mask(shape_of(f)); }

}  // namespace puyo

File: puyo/controller.py
"""操作列生成（組ぷよを実際のキー入力で目的の位置まで運ぶ）。

組ぷよの状態は「軸ぷよの位置 (x, y) と子ぷよの向き r」。r: 0=上, 1=右, 2=下, 3=左。
出現時は軸 (3, 12)・子 (3, 13)・r=0。移動中は 14 段目まで入れる（15 段目以上は天井）。
14 段目に残っているぷよのマスは通れない。

キー:
  L / R : 左右に 1 列。軸と子の行き先が両方空いていること
  A / B : 右回転 / 左回転。子の行き先が塞がっていたら蹴る:
          - 横向き（r=1, 3）になれない → 軸が反対側へ 1 列ずれる（壁蹴り）
          - 下向き（r=2）になれない → 軸が 1 段上がる（床蹴り。繰り返すと「回し」で高い列を登れる）
  Q     : クイックターン（回転キー 2 回）。左右とも塞がって回転できないとき、上下を入れ替える（必要なら軸が 1 段上がる）
最後に真下へ落として置く。

フレーム数（書籍の値を参考）: キー 1 回 2f、クイックターン 4f、落下 1 段 2f、接地 20f、
ちぎり +20f ＋ 離れたぷよの落下（1 段 10f、以降 1 段ごとに +6f）。

探索はスタート位置からのダイクストラ法（状態は 6 列 × 14 段 × 4 向き = 336 通り）。
AI の合法手（どこに置けるか）もこの結果に揃える。
"""

from __future__ import annotations

import heapq
from dataclasses import dataclass

from .core import ALL_MOVES, WIDTH, Move

MAX_ROW = 14  # 移動中に入れる最も高い段
SPAWN = (3, 12, 0)
DX = (0, 1, 0, -1)
DY = (1, 0, -1, 0)

KEY_FRAMES = 2
QUICK_TURN_FRAMES = 4
DROP_FRAMES_PER_ROW = 2
LOCK_FRAMES = 20
TEAR_FRAMES = 20


def tear_fall_frames(rows: int) -> int:
    """ちぎれたぷよが rows 段落ちるのにかかるフレーム（1 段 10f、以降 +6f ずつ）。"""
    return 0 if rows <= 0 else 10 + 6 * (rows - 1)


@dataclass(frozen=True)
class Operation:
    move: Move
    keys: str  # 例 "LLA"（最後の落下は含まない）
    frames: int  # キー入力＋落下＋接地（＋ちぎり）

    def pretty(self) -> str:
        return pretty_keys(self.keys)


def pretty_keys(keys: str) -> str:
    table = {"L": "←", "R": "→", "A": "↻", "B": "↺", "Q": "⇅"}
    return "".join(table[k] for k in keys) + "↓"


class Controller:
    """盤面の高さから、置ける場所・キー列・フレーム数を求める。"""

    def __init__(self, heights: list[int], top: list[bool] | None = None):
        self.h = heights  # 列 1..6 の高さ（heights[0] が 1 列目。13 段目まで）
        self.top = top or [False] * WIDTH  # 14 段目にぷよが残っているか

    def free(self, x: int, y: int) -> bool:
        if not (1 <= x <= WIDTH and 1 <= y <= MAX_ROW and y > self.h[x - 1]):
            return False
        return not (y == MAX_ROW and self.top[x - 1])

    def valid(self, x: int, y: int, r: int) -> bool:
        return self.free(x, y) and self.free(x + DX[r], y + DY[r])

    def rotate(self, x: int, y: int, r: int, d: int):
        """d=+1 で右回転、-1 で左回転。回れなければ None。"""
        r2 = (r + d) % 4
        if self.valid(x, y, r2):
            return x, y, r2
        if r2 in (1, 3):  # 壁蹴り: 軸を反対側へ
            x2 = x - DX[r2]
            if self.valid(x2, y, r2):
                return x2, y, r2
        elif r2 == 2:  # 床蹴り: 軸を 1 段上へ
            if self.valid(x, y + 1, r2):
                return x, y + 1, r2
        return None

    def quick_turn(self, x: int, y: int, r: int):
        if r not in (0, 2) or self.rotate(x, y, r, 1) or self.rotate(x, y, r, -1):
            return None
        r2 = (r + 2) % 4
        if self.valid(x, y, r2):
            return x, y, r2
        if r2 == 2 and self.valid(x, y + 1, r2):
            return x, y + 1, r2
        return None

    def neighbors(self, s):
        x, y, r = s
        if self.valid(x - 1, y, r):
            yield "L", KEY_FRAMES, (x - 1, y, r)
        if self.valid(x + 1, y, r):
            yield "R", KEY_FRAMES, (x + 1, y, r)
        for key, d in (("A", 1), ("B", -1)):
            t = self.rotate(x, y, r, d)
            if t:
                yield key, KEY_FRAMES, t
        t = self.quick_turn(x, y, r)
        if t:
            yield "Q", QUICK_TURN_FRAMES, t

    def apply(self, keys: str, start=SPAWN):
        """キー列をなぞった後の状態。途中で動けなければ None。"""
        path = self.path(keys, start)
        return path[-1] if path else None

    def path(self, keys: str, start=SPAWN) -> list[tuple[int, int, int]] | None:
        """キー列をなぞった途中の状態（出現位置を含む）。途中で動けなければ None。"""
        s, out = start, [start]
        for k in keys:
            for key, _, t in self.neighbors(s):
                if key == k:
                    s = t
                    break
            else:
                return None
            out.append(s)
        return out

    def drop_frames(self, x: int, y: int, r: int) -> int:
        """(x, y, r) から真下に落として置くまでのフレーム（落下＋接地＋ちぎり）。"""
        cx = x + DX[r]
        if cx == x:
            low = min(y, y + DY[r])  # 下にあるぷよ
            fall = low - (self.h[x - 1] + 1)
            return fall * DROP_FRAMES_PER_ROW + LOCK_FRAMES
        ha, hc = self.h[x - 1], self.h[cx - 1]
        fall = y - (max(ha, hc) + 1)
        tear = abs(ha - hc)
        extra = TEAR_FRAMES + tear_fall_frames(tear) if tear else 0
        return fall * DROP_FRAMES_PER_ROW + LOCK_FRAMES + extra

    def plan(self) -> dict[Move, Operation]:
        """置ける場所ごとの最短の操作。出現位置が塞がっていれば空。"""
        if not self.valid(*SPAWN):
            return {}
        dist = {SPAWN: 0}
        keys = {SPAWN: ""}
        heap = [(0, SPAWN)]
        while heap:
            d, s = heapq.heappop(heap)
            if d > dist[s]:
                continue
            for key, cost, t in self.neighbors(s):
                nd = d + cost
                if nd < dist.get(t, 1 << 30) or (nd == dist.get(t) and keys[s] + key < keys[t]):
                    dist[t] = nd
                    keys[t] = keys[s] + key
                    heapq.heappush(heap, (nd, t))
        best: dict[Move, Operation] = {}
        for (x, y, r), d in dist.items():
            m = Move(x, r)
            total = d + self.drop_frames(x, y, r)
            cur = best.get(m)
            if cur is None or (total, keys[(x, y, r)]) < (cur.frames, cur.keys):
                best[m] = Operation(m, keys[(x, y, r)], total)
        return {m: best[m] for m in ALL_MOVES if m in best}


def _top(field) -> list[bool]:
    return [t != 0 for t in field.top]


def plan_operations(field) -> dict[Move, Operation]:
    """置ける場所ごとの最短の操作。C++ 版があればそれを使う（結果は同じ）。"""
    h, top = field.heights(), _top(field)
    try:
        from . import _puyocpp
    except ImportError:
        return Controller(h, top).plan()
    return {Move(x, r): Operation(Move(x, r), keys, frames) for x, r, keys, frames in _puyocpp.plan_operations(h, top)}


def operation_path(field, op: Operation) -> list[tuple[int, int, int]]:
    """操作の途中の組ぷよの状態（軸の x, y と向き r）。"""
    return Controller(field.heights(), _top(field)).path(op.keys)


def reachable_moves(field) -> set[Move]:
    h = field.heights()
    if max(h) <= 11:  # 12 段以上の列が無ければどこでも置ける
        return set(ALL_MOVES)
    return set(Controller(h, _top(field)).plan())
